### ISSUE TITLE
Error in density.default(Slopes[useg]): 'x' contains missing values

### ISSUE THREAD

**Issue by davismcc** [rhondabacher/SCnorm#9]
Hi Rhonda

I have a reasonably complicated dataset with cell types and batch structure. I am trying to run `SCnorm` with condition as "day" of the timecourse experiment. After QC and selection of highly-variable genes I am running SCnorm on ~4600 cells and 2500 genes.

After running for quite a while, SCnorm eventually fails with the error shown below. (sessionInfo also)

Do you know why this might be the case, and any ideas for getting this to work? I could share the dataset privately if that would help troubleshoot.

Best
Davis
 
PS: Congratulations on the publication in Nat. Methods!

```
> scnorm_data_hvg_cond <- SCnorm(counts(sce_hvg), Conditions = sce_hvg$day, OutputName = "SCNormalizedData", SavePDF = FALSE, reportSF = TRUE, NCores = 6)
[...]
Trying K = 4
Plotting count-depth relationship for normalized data...
Error in density.default(Slopes[useg]): 'x' contains missing values
Traceback:

1. SCnorm(counts(sce_hvg), Conditions = sce_hvg$day, OutputName = "SCNormalizedData", 
 .     SavePDF = FALSE, reportSF = TRUE, NCores = 6)
2. checkCountDepth(Data = Data, NormalizedData = NORMDATA, Conditions = Conditions, 
 .     OutputName = paste0(OutputName, "_SCnorm_count-depth-relationship.pdf"), 
 .     SavePDF = SavePDF, Tau = Tau, FilterCellProportion = FilterCellProportion, 
 .     FilterExpression = FilterExpression, NCores = NCores, ditherCounts = ditherCounts)
3. lapply(1:length(Levels), function(x) {
 .     initialEvalPlot(MedExpr = MedExprList[[x]][GeneFilterList[[x]]], 
 .         SeqDepth = SeqDepthList[[x]], Slopes = SlopesList[[x]], 
 .         Name = Levels[[x]], NumExpressionGroups, BeforeNorm = BeforeNorm)
 . })
4. FUN(X[[i]], ...)
5. initialEvalPlot(MedExpr = MedExprList[[x]][GeneFilterList[[x]]], 
 .     SeqDepth = SeqDepthList[[x]], Slopes = SlopesList[[x]], Name = Levels[[x]], 
 .     NumExpressionGroups, BeforeNorm = BeforeNorm)
6. density(Slopes[useg])
7. density(Slopes[useg])
8. density.default(Slopes[useg])
9. stop("'x' contains missing values")
```

```
> sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux Server 7.2 (Maipo)

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] SCnorm_1.1.0        reshape2_1.4.2      moments_0.14       
 [4] cluster_2.0.6       quantreg_5.29       SparseM_1.76       
 [7] future_1.4.0        DT_0.2              scran_1.2.2        
[10] BiocParallel_1.8.1  scater_1.3.47       Biobase_2.34.0     
[13] BiocGenerics_0.20.0 readr_1.1.0         dplyr_0.5.0        
[16] magrittr_1.5        cowplot_0.7.0       ggbeeswarm_0.5.3   
[19] ggthemes_3.4.0      viridis_0.4.0       viridisLite_0.2.0  
[22] ggplot2_2.2.1       knitr_1.15.1       

loaded via a namespace (and not attached):
  [1] uuid_0.1-2            backports_1.0.5       Hmisc_4.0-2          
  [4] RcppEigen_0.3.2.9.1   plyr_1.8.4            igraph_1.0.1         
  [7] repr_0.10             lazyeval_0.2.0        sp_1.2-4             
 [10] shinydashboard_0.5.3  splines_3.3.2         listenv_0.6.0        
 [13] digest_0.6.12         htmltools_0.3.5       checkmate_1.8.2      
 [16] memoise_1.0.0         limma_3.30.13         globals_0.9.0        
 [19] matrixStats_0.52.1    xts_0.9-7             colorspace_1.3-1     
 [22] crayon_1.3.2          RCurl_1.95-4.8        jsonlite_1.3         
 [25] tximport_1.2.0        lme4_1.1-12           survival_2.41-3      
 [28] zoo_1.7-14            gtable_0.2.0          zlibbioc_1.20.0      
 [31] MatrixModels_0.4-1    car_2.1-4             DEoptimR_1.0-8       
 [34] VIM_4.6.0             scales_0.4.1          DBI_0.6-1            
 [37] edgeR_3.16.5          Rcpp_0.12.8           xtable_1.8-2         
 [40] laeken_0.4.6          htmlTable_1.9         foreign_0.8-67       
 [43] proxy_0.4-17          Formula_1.2-1         stats4_3.3.2         
 [46] vcd_1.4-3             htmlwidgets_0.8       FNN_1.1              
 [49] RColorBrewer_1.1-2    acepack_1.4.1         XML_3.98-1.5         
 [52] nnet_7.3-12           locfit_1.5-9.1        dynamicTreeCut_1.63-1
 [55] labeling_0.3          AnnotationDbi_1.36.2  munsell_0.4.3        
 [58] tools_3.3.2           RSQLite_1.1           evaluate_0.10        
 [61] stringr_1.2.0         robustbase_0.92-7     nlme_3.1-131         
 [64] mime_0.5              biomaRt_2.30.0        pbkrtest_0.4-7       
 [67] beeswarm_0.2.3        e1071_1.6-8           smoother_1.1         
 [70] tibble_1.3.0          statmod_1.4.29        stringi_1.1.3        
 [73] lattice_0.20-35       IRdisplay_0.4.4       Matrix_1.2-8         
 [76] nloptr_1.0.4          lmtest_0.9-35         data.table_1.9.8     
 [79] bitops_1.0-6          httpuv_1.3.3          R6_2.2.0             
 [82] latticeExtra_0.6-28   gridExtra_2.2.1       vipor_0.4.5          
 [85] IRanges_2.8.1         codetools_0.2-15      boot_1.3-18          
 [88] MASS_7.3-45           assertthat_0.1        destiny_2.0.8        
 [91] rhdf5_2.18.0          rjson_0.2.15          S4Vectors_0.12.0     
 [94] mgcv_1.8-17           hms_0.3               grid_3.3.2           
 [97] rpart_4.1-10          IRkernel_0.7.1        class_7.3-14         
[100] minqa_1.2.4           Rtsne_0.11            TTR_0.23-1           
[103] pbdZMQ_0.2-5          scatterplot3d_0.3-39  shiny_1.0.1          
[106] base64enc_0.1-3 
```

**Comment by rhondabacher** [comment]
Hi Davis,

I checked where the error is occuring and have added some code so that it should no longer cause an error message. My best guess is that one of the per-gene estimated slopes returned as NA, but I haven't see that happen often. I just spent some time trying ways to force this to happen but was not able to reproduce it. If you are able to share the dataset then that would be great. I can learn what might cause the NA and implement a more complete solution. You can email me at rhonda.bacher@gmail.com

Thanks!
Rhonda

**Comment by rhondabacher** [comment]
I figured out the line that caused this and have permanently fixed it in v1.1.3.
